Add format option to `cargo tree` to print the package version requirement

devodev opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/cargo/pull/16192)*

### Description

- Add a new formatting option `{ver-req}` to `cargo tree` that prints the resolved version requirement of packages as specified in Cargo.toml.
- Add a unit test to verify that the version constraint is printed correctly.
- Update the `cargo tree` command documentation with the new option.

As per the issue linked below, I implemented the second suggestion which was to add a new formatting option.
Adding a new flag would also be relatively simple, but I'd like to know what people think before adding a commit.

Fixes: #16163

### Example

Example output using the cargo repo:

<details>

<summary>cargo tree output</summary>

```shell-session
$ cargo tree --format '{p} satisfies {vr}'
cargo v0.93.0 (E:\dev\github.com\rust-lang\cargo) satisfies *
├── annotate-snippets v0.12.4 satisfies ^0.12.4
│   ├── anstyle v1.0.13 satisfies ^1.0.4
│   ├── memchr v2.7.6 satisfies ^2.7.4
│   └── unicode-width v0.2.1 satisfies ^0.2.0
├── anstream v0.6.20 satisfies ^0.6.20
│   ├── anstyle v1.0.13 satisfies ^1.0.0
│   ├── anstyle-parse v0.2.7 satisfies ^0.2.0
│   │   └── utf8parse v0.2.2 satisfies ^0.2.1
│   ├── anstyle-query v1.1.4 satisfies ^1.0.0
│   │   └── windows-sys v0.60.2 satisfies ^0.60.0
│   │       └── windows-targets v0.53.4 satisfies ^0.53.2
│   │           └── windows_x86_64_msvc v0.53.0 satisfies ^0.53.0
│   ├── anstyle-wincon v3.0.10 satisfies ^3.0.5
│   │   ├── anstyle v1.0.13 satisfies ^1.0.0
│   │   ├── once_cell_polyfill v1.70.1 satisfies ^1.56.0
│   │   └── windows-sys v0.60.2 satisfies ^0.60.0 (*)
│   ├── colorchoice v1.0.4 satisfies ^1.0.0
│   ├── is_terminal_polyfill v1.70.1 satisfies ^1.48
│   └── utf8parse v0.2.2 satisfies ^0.2.1
├── anstyle v1.0.13 satisfies ^1.0.13
├── anyhow v1.0.100 satisfies ^1.0.100
├── base64 v0.22.1 satisfies ^0.22.1
├── blake3 v1.8.2 satisfies ^1.8.2
│   ├── arrayref v0.3.9 satisfies ^0.3.5
│   ├── arrayvec v0.7.6 satisfies ^0.7.4
│   ├── cfg-if v1.0.3 satisfies ^1.0.0
│   └── constant_time_eq v0.3.1 satisfies ^0.3.1
│   [build-dependencies]
│   └── cc v1.2.39 satisfies ^1.1.12
│       ├── find-msvc-tools v0.1.2 satisfies ^0.1.2
│       ├── jobserver v0.1.34 satisfies ^0.1.30
│       │   └── getrandom v0.3.3 satisfies ^0.3.2
│       │       └── cfg-if v1.0.3 satisfies ^1
│       └── shlex v1.3.0 satisfies ^1.3.0
├── cargo-credential v0.4.9 (E:\dev\github.com\rust-lang\cargo\credential\cargo-credential) satisfies ^0.4.2
│   ├── anyhow v1.0.100 satisfies ^1.0.100
│   ├── serde v1.0.228 satisfies ^1.0.228
│   │   ├── serde_core v1.0.228 satisfies =1.0.228
│   │   └── serde_derive v1.0.228 (proc-macro) satisfies ^1
│   │       ├── proc-macro2 v1.0.101 satisfies ^1.0.74
│   │       │   └── unicode-ident v1.0.20 satisfies ^1.0
│   │       ├── quote v1.0.41 satisfies ^1.0.35
│   │       │   └── proc-macro2 v1.0.101 satisfies ^1.0.80
│   │       │       └── unicode-ident v1.0.20 satisfies ^1.0
│   │       └── syn v2.0.106 satisfies ^2.0.81
│   │           ├── proc-macro2 v1.0.101 satisfies ^1.0.91
│   │           │   └── unicode-ident v1.0.20 satisfies ^1.0
│   │           ├── quote v1.0.41 satisfies ^1.0.35 (*)
│   │           └── unicode-ident v1.0.20 satisfies ^1
```

</details>

epage commented on src/cargo/ops/tree/graph.rs:
```diff
@@ -49,14 +50,34 @@ impl std::hash::Hash for NodeId {
+            .cmp(&other.package_id)
+            .then_with(|| self.features.cmp(&other.features))
+            .then_with(|| self.kind.cmp(&other.kind))
+        // version_req is not orderable
```
This is making Ord and Eq not match

epage commented on tests/testsuite/cargo_tree/deps.rs:
```diff
@@ -57,7 +57,8 @@ foo v0.1.0 ([ROOT]/foo)
 └── devdep v1.0.0
-    └── b v1.0.0 (*)
+    └── b v1.0.0
+        └── c v1.0.0
```
This seems unrelated, why did it change?

devodev replied:
I think you put the finger on it and it's probably caused by the Eq/Ord mismatch on NodePackage, testing it out now :)

epage commented on tests/testsuite/cargo_tree/deps.rs:
```diff
@@ -1158,7 +1182,7 @@ fn format() {
             dep = {version="1.0", optional=true}
             other-dep = {version="1.0", optional=true}
-            dep_that_is_awesome = {version="1.0", optional=true}
+            dep_that_is_awesome = {version=">=1.0, <2", optional=true}
```
please make test edits in their own commit. Even adding a place holder for the assert, just not yet using `{c}` yet. This commit would add that format option and the diff would be focused on just that and how the output changed.

devodev replied:
No problem, I'll split into commits.

devodev commented on tests/testsuite/cargo_tree/deps.rs:
```diff
@@ -1208,6 +1232,17 @@ foo v0.1.0 ([ROOT]/foo) MIT https://github.com/rust-lang/cargo
+        .with_stdout_data(str![[r#"
+foo v0.1.0 ([ROOT]/foo) satisfies *
+├── dep v1.0.0 satisfies ^1.0
+├── dep_that_is_awesome v1.0.0 satisfies >=1.0, <2
```
Let me add some more tests for other use-cases

devodev replied:
Updated tests, but can you tell me if this is what you expected to see?

devodev committed:
Update cargo tree command docs with new version requirement option

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
Update cargo tree format tests with version requirement scenarios

Update cargo tree format tests and introduce different version requirements
and new placeholder tests to prepare for the next commits which will add
a new formatting option to display package version requirements.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
Add version requirement format option to cargo tree

Also update unit tests with test cases for normal and inverted
output.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

epage commented on tests/testsuite/cargo_tree/deps.rs:
```diff
@@ -1246,25 +1246,25 @@ foo v0.1.0 ([ROOT]/foo) [bar,default,dep_that_is_awesome,foo,other-dep]
+├── dep v1.0.0 satisfies =1.0
+├── dep_that_is_awesome v1.0.0 satisfies >=1.0, <2
+└── other-dep v1.0.0 satisfies ^1.0
+    └── dep v1.0.0 satisfies =1.0
```
both paths to `dep` involve the same version requirement.  It would help if they were different.

devodev commented on src/cargo/ops/tree/graph.rs:
```diff
@@ -49,14 +50,34 @@ impl std::hash::Hash for NodeId {
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct NodePackage {
```
Interesting. Alright I can take a look at implementing Ord on `OptVersionReq` in a way that makes the most sense.

> Also, best if splitting something out like this be left to a refactor commit before the main commit.

>> FYI the refactor commit I was looking for was to have no behavior or logic changes, which means it wouldn't have the version requirement.

Sounds good, will move the version requirement changes to the next commit where we update the tests.

devodev replied:
I pushed an alternate implementation that does not require storing additional data on the Node itself.
Instead, we record the version requirement in the Graph mapped by PackageID. We have already other data recorded the same way so I thought it was fitting for this use-case as well.

One thing I am not sure is whether we should default to `OptVersionReq::Any` for packages without version requirement (root packages). This was the behavior I went with previously, but looking back now, I noticed that some format args like `license` conditionally write a value, therefore I replicated here as well. Only thing is if we use a format string like the one in tests with `{p} satisfies {ver-req}`, it shows as: `dep satisfies ` which doesn't feel great.

I'd like to know what people think of this. I personally am divided 😄

ehuss commented on src/cargo/ops/tree/format/mod.rs:
```diff
@@ -30,6 +33,7 @@ impl Pattern {
                 RawChunk::Argument("r") => Chunk::Repository,
                 RawChunk::Argument("f") => Chunk::Features,
                 RawChunk::Argument("lib") => Chunk::LibName,
+                RawChunk::Argument("c") => Chunk::Constraint,
```
I don't know. I assume it is because shorter forms are more convenient. One would need to look through the history in the original repo to see if that was ever mentioned.

I think using longer forms makes sense, with shortcuts only for common options.

epage replied:
For myself, I tend to shorten "version requirement" to "version-req" but I could see us going in a lot of different directions.

devodev committed:
Support hyphens in cargo tree format parser

devodev committed:
Add unit tests for cargo tree format parser

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

epage commented on tests/testsuite/cargo_tree/deps.rs:
```diff
@@ -1156,9 +1163,9 @@ fn format() {
 
             [dependencies]
-            dep = {version="1.0", optional=true}
+            dep = {version="=1.0"}
```
This is removing coverage.

I suspect the test changes here should be purely additive rather than mutating existing stuff.
